cpp/include/cudf/lists/set_operations.hpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+namespace cudf::lists {
+/**
+ * @addtogroup set_operations
+ * @{
+ * @file
+ */
+
+/**
+ * @brief Check if lists at each row of the given lists columns overlap.
+ *
+ * Given two input lists columns, each list row in one column is checked if it has any common
+ * elements with the corresponding row of the other column.
+ *
+ * A null input row in any of the input lists columns will result in a null output row.
+ *
+ * @throw cudf::logic_error if the input lists columns have different sizes.
+ * @throw cudf::logic_error if children of the input lists columns have different data types.
+ *
+ * Example:
+ * @code{.pseudo}
+ * lhs    = { {0, 1, 2}, {1, 2, 3}, null,         {4, null, 5} }
+ * rhs    = { {1, 2, 3}, {4, 5},    {null, 7, 8}, {null, null} }
+ * result = { true, false, null, false }
+ * @endcode
+ *
+ * @param lhs The input lists column for one side
+ * @param rhs The input lists column for the other side
+ * @param nulls_equal Flag to specify whether null elements should be considered as equal, default
+ *        to be `UNEQUAL` which means only non-null elements are checked for overlapping
+ * @param nans_equal Flag to specify whether floating-point NaNs should be considered as equal
+ * @param mr Device memory resource used to allocate the returned object
+ * @return A column of type BOOL containing the check results
+ */
+std::unique_ptr<column> have_overlap(
+  lists_column_view const& lhs,
+  lists_column_view const& rhs,
+  null_equality nulls_equal           = null_equality::UNEQUAL,
+  nan_equality nans_equal             = nan_equality::ALL_EQUAL,
```

Comment:
Issue #11329 filed.

Reply:
This is the special behavior of the `have_overlap`/`arrays_overlap`. When checking if two lists overlap, nulls are excluded (i.e., compared as unequal).

This behavior is implemented in Spark and in Presto, but not in Snowflake.

On the other hand, when finding union/intersect/except subset between two lists, nulls are compared as equal.

Reply:
To reduce surprises when the users call our APIs with default parameters, we should keep these default values. Otherwise, users must provide explicit nulls and nans handling parameters every time they call these APIs from the public interfaces.

Reply:
My expectation is that two lists have overlap if (and only if) their intersection is not empty. This function’s behavior should be self-consistent with the intersection behavior, which would require equal nulls. If Spark needs to deviate from that, it can supply the argument explicitly.

Reply:
That's possible. Then the check overlap API with default parameters will be closely following Snowflake `arrays_overlap` instead.

FYI: Spark's `arrays_overlap` behavior is even more complicated to achieve, not just by changing parameters. I had to add specialized code to get it: https://github.com/rapidsai/cudf/pull/11143/files#diff-b16ca9906273a04b92cc1ab075914909f7f852407d4135adb9e32049b38e9cf0R101